Add SFX for multiplayer lobby countdown timer

nekodex opened the pull request:
Currently plays a tick every second when there are 10 seconds (or less) remaining, with a more obnoxious sample layered on top for the last 3 seconds... open to suggestions for better cutoff values.

I'm also not sure if `MultiplayerReadyButton` is the best place to implement this as I noticed it doesn't seem to update at a consistent interval?

- [x] depends on ppy/osu-resources#184

nekodex committed:
Add multiplayer lobby countdown SFX

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs:
```diff
@@ -83,6 +96,16 @@ private void updateButtonText()
                 else
                     countdownRemaining = countdown.TimeRemaining - timeElapsed;
 
+                if (countdownRemaining.Seconds <= 10 && (lastTickPlayed == null || lastTickPlayed != countdownRemaining.Seconds))
```
This code should probably not be in a method named `updateButtonText()`. At the least I'd do a rename of the method here to something more generic, because I'm not super sure it's worth pulling this out of this method into another...

peppy replied:
Will restructure this on top of https://github.com/ppy/osu/pull/17643, which creates a place this code can be executed.

bdach commented on osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs:
```diff
@@ -83,6 +96,16 @@ private void updateButtonText()
+                {
+                    countdownTickSample?.Play();
+
+                    if (countdownRemaining.Seconds <= 3)
+                        countdownTickFinalSample?.Play();
```
This sample plays on the counts of 3, 2, 1 and 0, and I was half expecting the 0 count sample to be somehow different, e.g. higher in pitch, to signal that it's the end of the countdown, but it's not so, it plays exactly the same which feels... off? Maybe I've been conditioned to expect this by other games but yeah.

peppy replied:
I was also expecting a pitch shift, as it gets close to zero. As I mentioned on the resource PR I'm not a huge fan of the warning-bell style of this sound right now.

peppy replied:
Some feedback on the sounds:

![CleanShot 2022-04-04 at 05 19 33](https://user-images.githubusercontent.com/191335/161478736-f61825df-3357-4d5c-997a-f35b6d109e38.png)

peppy committed:
Update resources

peppy committed:
Merge branch 'master' into mp-countdown-sfx

peppy committed:
Update resources

peppy commented:
I've updated this in line with the changes I made in #17643. Still not 100% on the sounds but let's get this in a release for some wider feedback I guess?

bdach commented:
There's something weird happening with the sound if I put on a countdown and stay in the room without readying up myself. It just kinda... layers over itself multiple times one second past the zero count? See video (warning: _loud_)

https://user-images.githubusercontent.com/20418176/161594529-a4566356-5b6d-42e9-a1d9-9751ebd9128b.mp4

bdach reviewed (changes requested):
see above

bdach commented:
The following diff fixes both issues for me:

```diff
diff --git a/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs b/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs
index a9092bc25a..56828d61c1 100644
--- a/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs
+++ b/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs
@@ -72,6 +72,15 @@ private void scheduleNextCountdownUpdate()
                 // If a countdown is active, schedule relevant components to update on the next whole second.
                 double timeToNextSecond = countdownTimeRemaining.TotalMilliseconds % 1000;
 
+                // sometimes it happens that the next countdown update is scheduled just a fraction too early, and as such timeToNextSecond can be near zero,
+                // while it generally should be somewhere close to 1000ms (the exact definition of "sometimes" and magnitude of "how close" depends on the framerate and frame alignment).
+                // this will lead to the countdown update firing in two consecutive frames, which is undesirable.
+                // therefore, assume that the next frame update will take the same amount of time as the previous one,
+                // and intentionally delay updates that are too close to the duration between frames.
+                // note while the estimate is based on past framerate, it is very conservative anyhow (timeToNextSecond really should be much larger than that).
+                if (timeToNextSecond < Time.Elapsed)
+                    timeToNextSecond += 1000;
+
                 countdownUpdateDelegate = Scheduler.AddDelayed(onCountdownTick, timeToNextSecond);
             }
             else
@@ -89,7 +98,8 @@ void onCountdownTick()
                 if (secondsRemaining < 10) countdownTickSample?.Play();
                 if (secondsRemaining <= 3) countdownTickFinalSample?.Play();
 
-                scheduleNextCountdownUpdate();
+                if (secondsRemaining > 0)
+                    scheduleNextCountdownUpdate();
             }
         }
 
```

The `secondsRemaining > 0` check is pretty safe I think, but the other thing is... pretty complicated. I haven't dug too deep into it, but here's a log excerpt that may help to make sense of this issue:

```
[runtime] 2022-04-04 17:36:47 [verbose]: scheduleNextCountdownUpdate: current time @ 20289.76788, time to next second is 999.4092999999993
[runtime] 2022-04-04 17:36:48 [verbose]: onCountdownTick fired @ 21290.243072999998
[runtime] 2022-04-04 17:36:48 [verbose]: scheduleNextCountdownUpdate: current time @ 21290.243072999998, time to next second is 0.6798999999991793
[runtime] 2022-04-04 17:36:48 [verbose]: onCountdownTick fired @ 21294.570649
[runtime] 2022-04-04 17:36:48 [verbose]: scheduleNextCountdownUpdate: current time @ 21294.570649, time to next second is 997.0380999999998
```

peppy commented:
I originally had something similar to your diff in my prereq PR, but I omitted it for simplicity (since it didn't affect the text display)..

peppy commented:
The reason I didn't go with it in the end is that it has the potential to completely skip one update, which we don't want. I'll try something else.

peppy commented:
The larger silly problem here is that this code is using realtime, not game time. I'm not sure why it is, but it causes issues all over the place due to time changing on the same update call.

peppy committed:
Fix multiple issues with timekeeping

- Using realtime (`DateTimeOffset.Now`) meant that values would be
  changing in the same frame, causing misfirings or incorrect displays
- No debounce on sample playback meant that scheduling edge cases could
  potentially cause samples to be played more than once.

peppy commented:
I've applied some fixes to bring things to order.

One annoying thing I haven't had the time to look into is that `TestSceneMatchStartControl` won't work after you click the ready button twice. @smoogipoo might be something you can make a quick fix for? I took a quick look but without debugging the full flow I'm not sure where to look.

peppy committed:
Disable "final" sample in countdown for the time being

peppy committed:
Appease the CI

smoogipoo commented:
> `TestSceneMatchStartControl` won't work after you click the ready button twice

This is because the second click enters gameplay.

smoogipoo commented:
Crashes when cancelling a 30 second countdown:
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at osu.Game.Screens.OnlinePlay.Multiplayer.Match.MultiplayerReadyButton.get_countdownTimeRemaining() in /home/smgi/Repos/osu/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs:line 167
   at osu.Game.Screens.OnlinePlay.Multiplayer.Match.MultiplayerReadyButton.<scheduleNextCountdownUpdate>g__onCountdownTick|19_0() in /home/smgi/Repos/osu/osu.Game/Screens/OnlinePlay/Multiplayer/Match/MultiplayerReadyButton.cs:line 87
```
The whole debouncing stuff is way over my head, with a time-modulo in the scheduling and then a `< 500` check in the schedule - it's a bit much and I hope it can be simplified. But it seems to work so... :see_no_evil:

peppy committed:
Full potential null reference and add better commentary on countdown scheduling

peppy committed:
Remove debounce logic (not required after switching to `Update` clock time)

peppy committed:
Reword comment slightly

peppy commented:
I've fixed this issue, and also removed the debounce logic. As I suspected, it is no longer required after switching to update clock time. I've tried to repro the double-playback issue as pointed out in https://github.com/ppy/osu/pull/17604#issuecomment-1087841184 but can't after many tries.
